clang/lib/Driver/Driver.cpp

```diff
@@ -5203,761 +5225,816 @@ class OffloadingActionBuilder final {
+          Action *FPGAAOTAction;
+          if (Input->getType() == types::TY_FPGA_AOCR ||
+              Input->getType() == types::TY_FPGA_AOCR_EMU)
+            // Generate AOCX/AOCR
+            FPGAAOTAction =
+                C.MakeAction<BackendCompileJobAction>(Input, FPGAOutType);
+          else if (Input->getType() == types::TY_FPGA_AOCX)
+            FPGAAOTAction = Input;
           else
-            continue;
-        } else if (A->getOption().matches(options::OPT_Xsycl_backend)) {
-          // Passing device args: -Xsycl-target-backend <opt>
-          TargetBE = &SYCLTripleList.front();
-          DeviceOpt = A->getValue(0);
-        } else if (A->getOption().matches(options::OPT_Xs_separate)) {
-          // Passing device args: -Xs <opt>
-          DeviceOpt = A->getValue(0);
-        } else {
+            llvm_unreachable("Unexpected FPGA input type.");
+          // Rename the column within the table.  The list of files that is
+          // produced from the AOT offline complation step is just a list of
+          // files.  Rename with the [Code] designator to be processed by
+          // the offload-wrapper in -batch mode.
+          auto *RenameAction = C.MakeAction<FileTableTformJobAction>(
+              FPGAAOTAction, types::TY_Tempfilelist, types::TY_Tempfilelist);
+          RenameAction->addRenameColumnTform(FileTableTformJobAction::COL_ZERO,
+                                             FileTableTformJobAction::COL_CODE);
+          auto *DeviceWrappingAction = C.MakeAction<OffloadWrapperJobAction>(
+              RenameAction, types::TY_Object);
+          addDeps(DeviceWrappingAction, TC, BoundArch);
           continue;
-        };
-      }
-      useNative = false;
-      if (needLibs)
-        if (TC->getTriple().getSubArch() == llvm::Triple::SPIRSubArch_gen &&
-            TargetOpt && DeviceOpt) {
-
-          auto checkBF = [=](std::string &Dev) {
-            static const std::regex BFFs("pvc.*|ats.*");
-            return std::regex_match(Dev, BFFs);
-          };
-
-          needLibs = true;
-          std::string Params{DeviceOpt};
-          size_t DevicesPos = Params.find("-device ");
-          useNative = false;
-          if (DevicesPos != std::string::npos) {
-            useNative = true;
-            std::istringstream Devices(Params.substr(DevicesPos + 8));
-            for (std::string S; std::getline(Devices, S, ',');) {
-              useNative &= checkBF(S);
-            }
+        } else if (!types::isFPGA(Input->getType())) {
+          // No need for any conversion if we are coming in from the
+          // clang-offload-deps or regular compilation path.
+          if (IsNVPTX || IsAMDGCN || ContainsOffloadDepsAction(Input) ||
+              ContainsCompileOrAssembleAction(Input)) {
+            LinkObjects.push_back(Input);
+            continue;
           }
+          Action *ConvertSPIRVAction = C.MakeAction<SpirvToIrWrapperJobAction>(
```

Comment:
Can you please add a short comment on how we will get here and why this transformation is needed?

Thanks